### ISSUE TITLE
Add MIPS

### ISSUE THREAD

**Issue by DonaldKellett** [codewars/codewars-runner-cli#819]
Please complete the following information about the language:

- Name: [MIPS](https://en.wikipedia.org/wiki/MIPS_architecture) assembly
- Language Version: [spim](http://spimsimulator.sourceforge.net) (the terminal interface)? But 1) it's a simulator so it might be slow and 2) we probably have to roll our own testing framework. Alternatively, we could look into [cross-compilation](https://stackoverflow.com/a/12482704) or similar to see if we can save effort on the testing framework by using [Criterion](https://github.com/Snaipe/Criterion).

The following are optional, but will help us add the language:

- Test Frameworks: Probably [MIPSUnit](https://www.researchgate.net/publication/314296181_MIPSUnit_A_Unit_Testing_Framework_for_MIPS_Assembly) (is it even open-source?) if we're using `spim`, or make [Criterion](https://github.com/Snaipe/Criterion) work with MIPS (like what we do for NASM)
- How to install: `sudo apt-get install spim` (?) ([source](https://askubuntu.com/a/367173))
- How to compile/run: `spim -file Solution.s` (where `Solution.s` is the solver's MIPS solution file) (?) - refer to the [Ubuntu manpages](http://manpages.ubuntu.com/manpages/cosmic/man1/spim.1.html) for more details
- Any comments: MIPS is a common choice of assembly language in introductory computer architecture courses due to its relatively simple instruction set.

---

:+1: reaction _might_ help.

**Comment by DonaldKellett** [comment]
_Note: Just putting this out there for now, might look into it further and provide additional details in the future if time allows. In the meantime, anyone more experienced in MIPS is welcome to contribute :smile:_

**Comment by DonaldKellett** [comment]
Found a possible approach to support testing: `MIPSUnit::MUnit`. It's a testing framework for [MARS](http://courses.missouristate.edu/kenvollmar/mars/) MIPS in Java based on JUnit.

Download: https://cis.gvsu.edu/~kurmasz/Software/mipsunit_munit/download.html - this downloads `munit.jar`. NOTE: separate installation of MARS MIPS is **not** required.

Usage: Just follow the example workflow in https://cis.gvsu.edu/~kurmasz/Software/mipsunit_munit/ - it's very easy to use

Challenge: How to make it output Codewars format, especially since the Java source code does not seem to be openly available? Maybe we can decompile the `.jar` and modify the resulting Java file to output Codewars format like what we do for JUnit. But I would recommend double-checking copyright / license terms for `MIPSUnit::MUnit` before doing so.

**Comment by DonaldKellett** [comment]
As for the language icon, since the official MIPS logo seems too wide, how about [a microchip with "MIPS" written on it](https://github.com/codewars/codewars-runner-cli/files/4677868/microchip-solid.svg.zip)? Note that this proposed logo is modified from the microchip logo [offered by Font Awesome](https://fontawesome.com/license/free).

**Comment by kurmasz** [comment]
Who wants access to the Java source for MUnit?  Just send me your github account name.

**Comment by kurmasz** [comment]
Also, let me know if it makes more sense to have a separate fork for CodeWars, or if you just want to add some code to the main distribution.

**Comment by DonaldKellett** [comment]
Thanks @kurmasz, my GitHub account name is @DonaldKellett :smile: 

@kazk Since I've got plenty of spare time today with my finals out of the way, how about I work on it in a private repo and transfer ownership to Codewars when I'm done? Or would you like to work on it on your own when time allows?

**Comment by kazk** [comment]
@DonaldKellett Go ahead. I don't think I'll be able to work on this any time soon.

I noticed https://cis.gvsu.edu/~kurmasz/Software/mipsunit_munit says it's JUnit based.
Maybe you can contribute a way to specify a test listener class. Then we should be able to use Codewars JUnit listener we already have and others can benefit from the change as well.

**Comment by DonaldKellett** [comment]
> Maybe you can contribute a way to specify a test listener class. Then we should be able to use Codewars JUnit listener we already have and others can benefit from the change as well.

Thanks for the tip, I'll look into it.

**Comment by kazk** [comment]
@DonaldKellett We use JUnit 5's test runner and that also supports running tests written in JUnit 4.

**Comment by kazk** [comment]
Java tests works like the following:

- Compile tests and solution
- Test runner finds the compiled tests and runs them

Is it possible to generate a standalone tests? So instead of running tests with `java -jar munit.jar in_range_correct.asm InRangeTest.class`, the name of the assembly file is provided at compile time somehow (or just always use `solution.asm`) and `javac -cp munit.jar InRangeTest.java` produces a regular JUnit test file that can be executed without `munit.jar`.
If that's possible, we shouldn't need to do anything extra to add support.

**Comment by DonaldKellett** [comment]
cc @kurmasz for the question above

**Comment by kurmasz** [comment]
Can we produce standalone tests?  It depends on exactly what you mean.  It is certainly not necessary to have `munit.jar` be the entry point; but, the Mars simulator code and the MUnit "glue" code have to go somewhere.

If the primary concern is to avoid having `munit.jar` be the entry point, I think all you would need to do is put the setup code from `MUnitCore.java` (lines 260 to 335) into a JUnit `@Before` method.  The first thing that comes to mind is to make a parent test class with this `@Before` method and then require that test files for CodeWars inherit from this class.

If going this route, there is no reason why this couldn't be done in JUnit 5. The only part of MUnit that is tied to JUnit 4 is the few lines of code in `MUnitCore` that run the tests and report the errors.

**Comment by kazk** [comment]
@DonaldKellett Can you invite me to the private repo? I'll take a look and see what's needed.

**Comment by DonaldKellett** [comment]
@kazk I have sent you an invitation, let me know in case you run into any issues.

**Comment by kazk** [comment]
@kurmasz Thanks. I just skimmed through `MUnitCore.java` and noticed it adds lots of useful messages for each exception from `MARSSimulator`. So I don't think using our existing listener for Java is a good idea since all of those will be lost.
But using `@Before` instead of `MUnitCore.java` might be a good idea for our use case and it's worth exploring.

Looks like we need to do:
1. Initialize `MARSSimulator` (`mars`)
2. Set `MUnit`'s simulator to `mars`
3. Load "solution.asm" and handle any errors
4. Assemble it and handle any errors
5. Run tests

@DonaldKellett Were you able to build the project?

For implementing Codewars support, I'd probably start by writing `CodewarsListener` class that accepts an instance of `MARSSimulator`.
Requirements:
- Reset the simulator before each test like how the anonymous `RunListener` does in `MUnitCore.java`
- Output the test results as it runs by implementing the listener methods
  - Add any additional information when the test fails by throwing something other than the `AssertionError` like how it's done in `MUnitCore.java`

Then We can just replace the anonymous `RunListener` and get the Codewars output.

Ideally, we can use JUnit 5 instead because its `TestExecutionListener` exposes more info and allow us to produce better output. But we can start from JUnit 4 and see how that goes.
Updating from JUnit 4's `RunListener` to JUnit 5's `TestExecutionListener` should be pretty easy so I don't think it's a waste to first make it work with JUnit 4.

**Comment by kurmasz** [comment]
Keep in mind that the purpose of the detailed exceptions from `MarsSimulator` is to help novice assembly programmers debug their programs.  You may not need that level of detail in the CodeWars context.  

I agree about the `@BeforeClass` as opposed to `@Before` for the setup. 

I've been thinking about upgrading MUnit to use JUnit5, I just haven't had the time.

**Comment by DonaldKellett** [comment]
Thanks for all the pointers, I'll take a detailed look at it over the weekend :+1:

**Comment by DonaldKellett** [comment]
@kurmasz How do I build MIPSUnit from source to obtain the `munit.jar`?

@kazk I've just taken a look at `MUnitCore.java` and think I've found [the anonymous `RunListener`](https://github.com/DonaldKellett/MIPSUnit-Codewars/blob/master/src/edu/gvsu/mipsunit/munit/MUnitCore.java#L370). IIUC all I have to do is implement a `CodewarsListener` class extending `RunListener`, implement the listener methods accordingly and replace [lines 370 through 375 in `MUnitCore.java`](https://github.com/DonaldKellett/MIPSUnit-Codewars/blob/master/src/edu/gvsu/mipsunit/munit/MUnitCore.java#L370) with an instance of `CodewarsListener`?

**Comment by kurmasz** [comment]
It uses ant.  In the directory containing "build.xml", run "ant munit" or "ant all"

**Comment by kazk** [comment]
@DonaldKellett Yeah, then remove https://github.com/DonaldKellett/MIPSUnit-Codewars/blob/master/src/edu/gvsu/mipsunit/munit/MUnitCore.java#L384-L474 . You can open a new issue in your repo to discuss further if you want.

**Comment by DonaldKellett** [comment]
@kazk I have implemented a prototype for the Codewars listener, feel free to review my PR over at MIPSUnit-Codewars and see if that is close to what is needed :smile:

**Comment by DonaldKellett** [comment]
@kazk Not sure if you missed my message above but anyway the pull request is available at https://github.com/DonaldKellett/MIPSUnit-Codewars/pull/1 and awaiting your review.

If you're currently busy with other stuff then feel free to tell me and maybe I'll ping you again in two weeks.

**Comment by kazk** [comment]
@DonaldKellett Sorry, I saw it but forgot about it. I just added a short review.
